core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java

```diff
@@ -0,0 +1,177 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.sql.async;
+
+import io.questdb.MessageBus;
+import io.questdb.cairo.sql.NetworkSqlExecutionCircuitBreaker;
+import io.questdb.cairo.sql.PageAddressCacheRecord;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreakerConfiguration;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.mp.Job;
+import io.questdb.mp.MCSequence;
+import io.questdb.mp.RingQueue;
+import io.questdb.std.Misc;
+import io.questdb.std.Rnd;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class PageFrameReduceJob implements Job, Closeable {
+
+    private final static Log LOG = LogFactory.getLog(PageFrameReduceJob.class);
+    private final PageAddressCacheRecord[] record;
+    private final int[] shards;
+    private final int shardCount;
+    private final MessageBus messageBus;
+    private final SqlExecutionCircuitBreaker[] circuitBreakers;
+
+    // Each thread should be assigned own instance of this job, making the code effectively
```

Comment:
AFAIU there is a single instance of this object created in `O3Utils#setupWorkerPool()` which means that all of the workers share the same shard walk sequence.